2_process/src/raster_per_PRMS.R

```diff
@@ -0,0 +1,94 @@
+rbind.fill.list <- function(x) {
+  nam <- sapply(x, names)
+  unam <- unique(unlist(nam))
+  len <- sapply(x, length)
+  out <- vector("list", length(len))
+  for (i in seq_along(len)) {
+    out[[i]] <- unname(x[[i]])[match(unam, nam[[i]])]
+  }
+  setNames(as.data.frame(do.call(rbind, out), stringsAsFactors=FALSE), unam)
+}
+
+raster_to_catchment_polygons <- function(polygon_sf, raster,
+                                         categorical_raster = NULL,
+                                         raster_summary_fun = NULL,
+                                         new_cols_prefix = NULL){
+
+  #' @description Function extracts raster pixel per catchment polygon and calculated proportion of raster class per total
+  #'
+  #' @param polygon_sf sf object of polygon(s) in aoi
+  #' @param raster path to '.tif' file or raster object of SpatRaster object  
+  #' @param categorical_raster logical. If categorical raster, TRUE
+  #' @param raster_summary_fun for continuous raster, function to summarize the data by geometry. kept NULL if categorical_raster = TRUE
+  #' @param categorical_raster_prefix for categorical raster, prefix added to colnames to label raster classes/values
+  #' @value A data frame the same columns as the polygon_sf spatial dataframe with additional columns of the Land Cover classes
+  #' @example land cover raster ex: raster_to_catchment_polygons(polygon_sf = p1_catchments_sf, raster = p1_backcasted_LC[1], categorical_raster = TRUE, new_cols_prefix = 'lc_)
+  #' @example cont. raster ex: raster_to_catchment_polygons(polygon_sf = p1_catchments_sf, raster = p1_rod_salt, categorical_raster = FALSE, raster_summary_fun = sum, new_cols_prefix = 'cont_raster')
+  
+  # read in 
+  raster <- rast(raster)
+  vector_sf <- polygon_sf 
+  
+  ## check vector geometries 
+  if(any(!st_is_valid(vector_sf))){
+    vector_sf <- st_make_valid(vector_sf)
+    print('shp geometries fixed')
+    } 
```

Comment:
I noticed that some of the PRMS HRU polygons have loops/holes in them (I don't know the technical term for this). Is that the problem that this function fixes?

Reply:
Yes trying to see what's happening here. Also becoming an issue for gridmet. seem new issue #54